### PR TITLE
226: support locale variants

### LINKED ISSUE #226: Locales: Add support for variants of a locale
Mentioned this some time ago in [#glotpress](https://wordpress.slack.com/archives/glotpress/p1436617754000108).

Some ideas on how we could store variants in [GP_Locales](https://github.com/GlotPress/GlotPress-WP/blob/1.0.0/locales/locales.php#L88):

1) Define new `GP_Locale` with variant in `slug`:

``` php
$de_formal = new GP_Locale();
$de_formal->english_name = 'German (Formal)';
$de_formal->native_name = 'Deutsch (Sie)';
$de_formal->lang_code_iso_639_1 = 'de';
$de_formal->country_code = 'de';
$de_formal->wp_locale = 'de_DE_formal';
$de_formal->slug = 'de/formal';
$de_formal->google_code = 'de';
$de_formal->facebook_locale = 'de_DE';
```

2) Clone existing `GP_Locale` and add `variant` as a new property:

``` php
$de_formal = clone $de;
$de_formal->english_name = 'German (Formal)';
$de_formal->native_name = 'Deutsch (Sie)';
$de_formal->wp_locale = 'de_DE_formal';
$de_formal->variant = 'formal';
```

3) Define `variants` as a new property of the existing `GP_Locale`

``` php
$de = new GP_Locale();
$de->english_name = 'German';
$de->native_name = 'Deutsch';
$de->lang_code_iso_639_1 = 'de';
$de->country_code = 'de';
$de->wp_locale = 'de_DE';
$de->slug = 'de';
$de->google_code = 'de';
$de->facebook_locale = 'de_DE';
$de->variants = array(
    'formal' => array(
        'english_name' => 'German (Formal)',
        'native_name' => 'Deutsch (Sie)',
    ),
);
```

What we're currently doing on api.w.org side: https://meta.trac.wordpress.org/browser/sites/trunk/api.wordpress.org/public_html/translations/lib.php?rev=2319&marks=54-68#L53

### PR DESCRIPTION
**Pull request by toolstack**
Resolves #226 .

### CODE REVIEW

File: locales/locales.php [hunk 1]
```diff
@@ -972,6 +1010,8 @@ public function __construct() {
 		$gsw->country_code = 'ch';
 		$gsw->wp_locale = 'gsw';
 		$gsw->slug = 'gsw';
+		$gsw->variant_root = $de->slug;
```
Comment (swissspidy) [inline locales/locales.php]: Please note that while German (Switzerland) as a variant of standard German makes sense, Swiss German (gsw) is something completely different. It's a dialect (well, more than 26 different dialects to be precise) that has nothing in common with regular German. As such, `$gsw` shouldn't be a variant of `$de`.

Comment (pixelverbieger) [reply locales/locales.php]: … while _de_DE_formal_ is missing and should be added to the list of variants for German

Comment (swissspidy) [reply locales/locales.php]: That would be handled somewhere else though 🙂 `de_DE_formal` is not listed in `locales.php` after all, it's probably just a WordPress.org thing.

Comment (toolstack) [reply locales/locales.php]: Thanks @swissspidy, I'll remove the relationship.

@pixelverbieger I'm sure there will be many new variants added, like de_DE_formal once the the PR is merged, currently swsisspidy is correct that the formal variants are handled by translate.w.org separately.

File: gp-templates/translation-row.php [hunk 2]
```diff
@@ -178,6 +178,32 @@
 			</dl>
 			-->
 
+			<?php if ( property_exists( $t, 'root_translation_set_id' ) ) : ?>
+			<dl>
+				<dt><?php _e( 'Root Translaton:', 'glotpress' ); ?></dt>
```
Comment (ocean90) [inline gp-templates/translation-row.php]: Typo

**Review by ocean90 — CHANGES_REQUESTED** [review]

File: gp-templates/translation-row.php (hunk 2 above)
Comment (toolstack) [reply gp-templates/translation-row.php]: Fixed.

**Comment by HughP** [comment]
I’m not sure why the code invokes iso 629-2 codes.  BCP 47 Tags should be used. 

I don’t have a problem with the fallbacks chosen. They make sense. But it seems that the coding seems more complex than needed. ISO 639-2 and 639-3 are alligned for the purposes of locale codes CLDR or Iana should have the already Approved codes.

**Comment by xavivars** [comment]
Any idea of progress on this PR? 

While discussing about the details could take forever (this language is/is-not a variant of another language), the functionality itself it's extremely useful for cases that are clear. For example, we now have *Valencian (ca-valencia, Catalan as spoken in Valencia), and falling back to English is simply not an option: there aren't enough translators to keep all themes/plugins up to date in both variants.

**Comment by toolstack** [comment]
It will be merged as part of 3.0, which is scheduled for a Dec 20th release.

**Comment by xavivars** [comment]
I guess ca-valencia will need to be added as a variant of ca in locales.php for this case to work, right?

**Comment by toolstack** [comment]
@xavivars Correct, it can be done for a site by either hooking in to the `gp_locale_definitions_arrary` filter or creating a custom locales file.

**Pull request merged**